### ISSUE TITLE
feat: add aria-describedby to the button and id attributes to the slots in fast-accordion-item

### ISSUE THREAD

**Issue by MaiLeeTR** [microsoft/fast#6491]
# 🙋 Feature Request

Add `aria-describedby` to the button and `id` attributes to the start and end slots to establish and announce a relationship between the heading text and the slots.

_VoiceOver on macOS was used for testing_

## 🤔 Expected Behavior

If `aria-describedby` is added to the button and references the `id` of the start and end slot, a user with screen reader software turned on will be conveyed the entire context of the heading. Without `aria-describedby` relationship, users are only conveyed the button text and have to perform additional steps to read the rest of the information in the heading element.

This solution will effectively convey all information to a user upon focus on the heading element.

## 😯 Current Behavior

**Current behavior with content in the end slot:**
When a user has VoiceOver turned on and uses the keyboard to initially focus on the heading element, the button content is announced to the user along with the directions for activating the button. The content in the end slot is not announced and the user will not know there is additional information present unless they perform additional steps.

First, VoiceOver announces heading to user:
<img width="859" alt="Screen Shot 2022-10-31 at 12 41 00 PM" src="https://user-images.githubusercontent.com/55507048/199074304-c6b0bd58-1ad8-4013-9d28-954f41ccd7e4.png">

Next, VoiceOver announces instructions to activate button and doesn't communicate additional information available in heading:
<img width="858" alt="Screen Shot 2022-10-31 at 12 48 01 PM" src="https://user-images.githubusercontent.com/55507048/199075044-b7dc7aee-9d14-4a2b-9156-e3c0457cc0bc.png">

## 💁 Possible Solution

Add the ability to add `aria-describedby` to the button and set `id` attributes on the start and end slots. This could be optional so consumers can choose whether or not to use this feature if it doesn't make sense for their implementation.

## 🔦 Context

This feature will help consumers in the use case that they would want all the information conveyed to a user on focus of the heading element. In our use case, we are implementing a badge in the end slot with a number and a screen reader text message.

## 💻 Examples

**How the feature will enhance current behavior:**
When a user has VoiceOver turned on and uses the keyboard to initially focus on the heading element, the heading and slot content will be read to the user. This will provide them with all the information available without the user having to perform additional navigation.

First, VoiceOver announces heading to user:
<img width="859" alt="Screen Shot 2022-10-31 at 12 44 05 PM" src="https://user-images.githubusercontent.com/55507048/199075205-2a41bfee-cbe0-48c1-83eb-06acf7c6e573.png">

Next, VoiceOver announces the screen reader text "1 unread report" and then the instructions to activate the button:
<img width="855" alt="Screen Shot 2022-10-31 at 12 44 33 PM" src="https://user-images.githubusercontent.com/55507048/199075536-20fcc94a-e793-42dc-95da-08c9b246e9ab.png">

**Comment by EisenbergEffect** [comment]
Awesome writeup @MaiLeeTR! Thank you 🙇 
/cc @chrisdholt for triage/investigation.

**Comment by chrisdholt** [comment]
I don't think we'll add id's specifically to _any_ slots because that would mean that the slots themselves need to be added into the document flow rather than simply projecting their children. While the scenario might solve this specific approach, it breaks the concept of slots where they by nature project their children rather than live in the a11y tree and dom themselves. Thanks for the great write-up, I'm not sure that there is a quick and easy fix to this, but happy to look into if there are better ways to enable composition which supports this. One of the key potential issues you'll hit with the web platform right now is that enabling this is issues with cross shadow root ARIA. If slotted elements have ID's those id's are in the light DOM and the button is in the shadow DOM (I assume this is why the request exists to apply those to slots).

I think that setting these as "default slotted content" might actually solve this as those get rendered into the shadow DOM rather than the light DOM. This would allow you to query the shadow root and set `aria-describedby` in connectedCallback to the id's of the default slotted content you provide. A larger alternative might be to fork the template if this is common enough.

**Comment by chrisdholt** [comment]
@MaiLeeTR just wanted to follow up and see if you had thoughts on the above response.

**Comment by MaiLeeTR** [comment]
Thanks for the explanation. Yes, the idea is that if slots can have ids, the button in the shadow DOM can reference them, but I see what you mean about that breaking the concept of slots. And adding ids to the children of slots won't work either since they're in the light DOM, and the button is in the shadow DOM.

Can you explain what you mean by "default slotted content"?

**Comment by chrisdholt** [comment]
> Thanks for the explanation. Yes, the idea is that if slots can have ids, the button in the shadow DOM can reference them, but I see what you mean about that breaking the concept of slots. And adding ids to the children of slots won't work either since they're in the light DOM, and the button is in the shadow DOM.
> 
> Can you explain what you mean by "default slotted content"?

With the exception of default slots, most slots such as start/end or the checked and indeterminate indicator in checkbox, support passing default content to those slots. The content is rendered _into_ the shadow DOM. I've pushed up a quick example to checkbox to illustrate the point: https://github.com/microsoft/fast/blob/e7d967a872c32225f91ec09e8b156c7431349ccd/packages/web-components/fast-foundation/src/checkbox/stories/checkbox.register.ts#L133-160

While less than ideal, this may help provide a method where the id can exist within the shadow root. The only issue at that point is setting the necessary aria reference on the button itself using JS at connected callback perhaps...

**Comment by MaiLeeTR** [comment]
@chrisdholt Thanks for the example. Our team may need to explore the accordion template more for use with badges. We're building these components as part of a design system for our consumers. We want our consumers to be able to easily customize the content in the end slot by only editing the HTML markup. It seems like adding badges as "default slotted content" would add more complexity for our consumers to customize the badge content without doing additional work. Currently, below is how we foresee our consumers adding the badge content. (We plan to refine this so that the accessibility text is a badge property.)

<img width="435" alt="Screen Shot 2023-01-05 at 10 44 16 AM" src="https://user-images.githubusercontent.com/55507048/210834208-85d81197-efda-4ef6-9adf-c6bf79a62f5e.png">

**Comment by janechu** [comment]
Unfortunately `@microsoft/fast-foundation` is being deprecated, refer to #6955. We will begin PR clean up soon which includes merging any contributions to the package and snap an archived branch for anyone to refer to, we will also be showcasing in our upcoming documentation site update any projects that are component libraries or libraries similar to Foundation that may fill that need if you need to migrate.
